### PR TITLE
Functional tests - Add test Enable and disable B2B mode

### PR DESCRIPTION
**Pull request by boubkerbribri**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Add functional test 'Enable and disable B2B mode and check it in front'
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | `TEST_PATH="functional/BO/13_shopParameters/04_customerSettings/01_customers/07_enableB2BMode.js" URL_FO=shopURL/ npm run specific-test`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18161)

### CODE REVIEW

**Review by NeOMakinG — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

File: tests/puppeteer/campaigns/functional/BO/13_shopParameters/04_customerSettings/01_customers/04_enableB2BMode.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        'testIdentifier',
+        `checkB2BMode${this.pageObjects.boBasePage.uppercaseFirstCharacter(test.args.action)}`,
+        baseContext,
+      );
+      page = await this.pageObjects.boBasePage.viewMyShop();
+      this.pageObjects = await init();
+      await this.pageObjects.foBasePage.changeLanguage('en');
+      await this.pageObjects.foBasePage.goToLoginPage();
+      await this.pageObjects.loginFOPage.goToCreateAccountPage();
+      const isBirthDateInputVisible = await this.pageObjects.loginFOPage.isCompanyInputVisible();
```
Comment (Progi1984) [inline tests/puppeteer/campaigns/functional/BO/13_shopParameters/04_customerSettings/01_customers/04_enableB2BMode.js]:
```suggestion
      const isCompanyInputVisible = await this.pageObjects.loginFOPage.isCompanyInputVisible();
```

Comment (boubkerbribri) [reply tests/puppeteer/campaigns/functional/BO/13_shopParameters/04_customerSettings/01_customers/04_enableB2BMode.js]: done thanks

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js [hunk 2]
```diff
@@ -59,9 +60,20 @@ module.exports = class customerSettings extends BOBasePage {
...
     await this.clickAndWaitForNavigation(this.saveGeneralFormButton);
     return this.getTextContent(this.alertSuccessBlock);
   }
+
+  /**
+   * Set B2B mode status
+   * @param toEnable
+   * @returns {Promise<string>}
+   */
+  async setB2BModeStatus(toEnable = true) {
```
Comment (PierreRambaud) [inline tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: This is the same code as setPartnerOfferStatus, only the selector is different. Maybe create a generic function with 
`this.enableB2BModeToggle` as parameter

**Review by NeOMakinG — DISMISSED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
One comment

**Review by NeOMakinG — COMMENTED** [review]
Waiting for answer of Pierre

File: tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js (hunk 2 above)
Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: we don't want to use or call a selector in tests files but I can't create a generic function.

Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: Fixed, WDYT @PierreRambaud @NeOMakinG ?

File: tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js [hunk 3]
```diff
@@ -14,53 +14,39 @@ module.exports = class customerSettings extends BOBasePage {
...
+        break;
+      case 'Email after registration':
+        await this.waitForSelectorAndClick(this.sendEmailAfterRegistrationLabel.replace('%TOGGLE', toEnable ? 1 : 0));
+        break;
+      case 'Redisplay cart at login':
+        await this.waitForSelectorAndClick(this.redisplayCartAtLoginLabel.replace('%TOGGLE', toEnable ? 1 : 0));
+        break;
+      default:
+        throw new Error(`${option} was not found in the page`);
+    }
```
Comment (NeOMakinG) [inline tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]:
```suggestion
  async setOptionStatus(selector, toEnable = true) {
    await this.waitForSelectorAndClick(selector.replace('%TOGGLE', toEnable ? 1 : 0));
```

Comment (NeOMakinG) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: And then replaces every use of .setOptionsStatus with selector instead, the use of a switch there is the perfect contrary of a reusable component method!

Don't forget to throw an error if waitForSelectorAndClick doesn't find the selector (not if you put a selector which isn't in a switch case)

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

File: tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js (hunk 3 above)
Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: as I said, we separate tests logic from pages logic, so we don't want to call the selector in test.

Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: The other option is to do it like this :

```js
  async setOptionStatus(selector, toEnable = true) {
    await this.waitForSelectorAndClick(selector.replace('%TOGGLE', toEnable ? 1 : 0));
    await this.clickAndWaitForNavigation(this.saveGeneralFormButton);
    return this.getTextContent(this.alertSuccessBlock);
  }
  
  setPartnerOfferStatus(toEnable = true) {
    return this.setOptionStatus(this.enablePartnerOfferLabel, toEnable);
  }
```

Comment (NeOMakinG) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: @PierreRambaud what do you think ?

Myself, I think thats a good compromise even if it adds unwanted complexity, but if they wants to separate logics, sometimes it's needed..

File: tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js [hunk 4]
```diff
@@ -21,58 +21,32 @@ module.exports = class customerSettings extends BOBasePage {
...
```
Comment (PierreRambaud) [inline tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]:
```suggestion
  async setOptionStatus(option, toEnable = true) {
    let selector;

    switch (option) {
    case OPTION_B2B:
        selector = this.enableB2BModeToggle;
        break;
    case OPTION_PARTNER_OFFER:
        selector = this.enablePartnerOfferLabel;
        break;
    case OPTION_BIRTH_DATE:
        selector = this.askForBirthDateLabel;
        break;
    case OPTION_EMAIL_REGISTRATION:
        selector = this.sendEmailAfterRegistrationLabel;
        break;
    case OPTION_CART_LOGIN:
        selector = this.redisplayCartAtLoginLabel;
        break;
    default:
        throw new Error(`${option} was not found in the page`);
    }

    await this.waitForSelectorAndClick(selector.replace('%TOGGLE', toEnable ? 1 : 0));
    await this.clickAndWaitForNavigation(this.saveGeneralFormButton);
    
    return this.getTextContent(this.alertSuccessBlock);
}


```

And should be call with:
```
await this.pageObjects.customerSettingsPage.setOptionStatus(options.OPTION_PARTNER_OFFER, test.args.enable);
```

Please change indentation, doing it with a scratch editor :sweat_smile:

Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/shopParameters/customerSettings/index.js]: The fix is done, WDYT @PierreRambaud @NeOMakinG

**Review by NeOMakinG — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Pull request merged**